src/ht.c

```diff
@@ -146,162 +149,198 @@ SDB_API void ht_free(SdbHt* ht) {
-		ht->freefn, ht->calcsizeK, ht->calcsizeV);
+		ht->freefn, ht->calcsizeK, ht->calcsizeV, ht->elem_size);
 
 	for (i = 0; i < ht->size; i++) {
-		if (!ht->table[i]) {
-			continue;
-		}
+		HtBucket *bt = &ht->table[i];
+		HtKv *kv;
+		ut32 j;
 
-		ht->table[i]->free = NULL;
-		ls_foreach_safe (ht->table[i], iter, tmp, kv) {
-			internal_ht_insert_kv (ht2, kv, false);
-			ls_delete (ht->table[i], iter);
+		BUCKET_FOREACH (ht, bt, j, kv) {
+			ht_insert_kv (ht2, kv, false);
 		}
 	}
 	// And now swap the internals.
 	swap = *ht;
 	*ht = *ht2;
 	*ht2 = swap;
+
+	ht2->freefn = NULL;
 	ht_free (ht2);
 }
 
-static bool internal_ht_insert_kv(SdbHt *ht, HtKv *kv, bool update) {
-	bool found = false;
-	if (!ht || !kv) {
-		return false;
-	}
-	ut32 bucket, hash = hashfn (ht, kv->key);
-	if (update) {
-		(void)ht_delete_internal (ht, kv->key, &hash);
-	} else {
-		(void)ht_find (ht, kv->key, &found);
+static void check_growing(SdbHt *ht) {
+	if (ht->count >= LOAD_FACTOR * ht->size) {
+		internal_ht_grow (ht);
 	}
-	if (update || !found) {
-		bucket = hash % ht->size;
-		if (!ht->table[bucket]) {
-			ht->table[bucket] = ls_newf ((SdbListFree)ht->freefn);
-		}
-		ls_prepend (ht->table[bucket], kv);
-		ht->count++;
-		// Check if we need to grow the table.
-		if (ht->count >= LOAD_FACTOR * ht->size) {
-			internal_ht_grow (ht);
+}
+
+static HtKv *reserve_kv(SdbHt *ht, const char *key, const int key_len, bool update) {
+	ut32 bucket = bucketfn (ht, key);
+	HtBucket *bt = &ht->table[bucket];
+	HtKv *kvtmp;
+	ut32 j;
+
+	BUCKET_FOREACH (ht, bt, j, kvtmp) {
+		if (is_kv_equal (ht, key, key_len, kvtmp)) {
+			if (update) {
+				freefn (ht, kvtmp);
+				return kvtmp;
+			}
+			return NULL;
 		}
-		return true;
 	}
-	return false;
+
+	HtKv *newkvarr = realloc (bt->arr, (bt->count + 1) * ht->elem_size);
+	if (!newkvarr) {
+		return NULL;
+	}
+
+	bt->arr = newkvarr;
+	bt->count++;
+	ht->count++;
+	return kv_at (ht, bt, bt->count - 1);
 }
 
-static bool internal_ht_insert(SdbHt* ht, bool update, const char* key, void* value) {
-	if (!ht) {
+bool ht_insert_kv(SdbHt *ht, HtKv *kv, bool update) {
+	HtKv *kv_dst = reserve_kv (ht, kv->key, kv->key_len, update);
+	if (!kv_dst) {
 		return false;
 	}
-	HtKv* kv = calloc (1, sizeof (HtKv));
-	if (kv) {
-		kv->key = dupkey (ht, key);
-		kv->value = dupval (ht, value);
-		kv->key_len = calcsize_key (ht, kv->key);
-		kv->value_len = calcsize_val (ht, kv->value);
-
-		if (!internal_ht_insert_kv (ht, kv, update)) {
-			freefn (ht, kv);
-			return false;
-		}
-		return true;
-	}
-	return false;
+
+	memcpy (kv_dst, kv, ht->elem_size);
+	check_growing (ht);
+	return true;
 }
 
-SDB_API bool ht_insert_kv(SdbHt *ht, HtKv *kv, bool update) {
-	return internal_ht_insert_kv (ht, kv, update);
+static bool insert_update(SdbHt *ht, const char *key, void *value, bool update) {
+	ut32 key_len = calcsize_key (ht, key);
+	HtKv* kv_dst = reserve_kv (ht, key, key_len, update);
```

Comment:
FIXME: this increments both bucket and ht counts, but it is not always right... If the element is being updated, and not inserted, the count should remain the same.

Reply:
actually this was already handled...